### PR TITLE
lrvick/deterministic builds

### PR DESCRIPTION
**Pull request by lrvick**
## Changes

Astifacts are now:
* tracked by Git LFS in the "dist" directory
* named according to standard architecture/os names
* built deterministically with a reproducible containerized toolchain
* signed/verifiable by multiple people

## Review flow

The review pattern for this PR (and future release PRs) is

1. Manually review any changes (particularly to src/ directory) 
2. ```make attest``` to reproduce binaries of current release from source code
3. ```make sign``` to attach your own signature to current release
4. Push your own signature to this release branch
5. Merge after 2+ signatures are aggregated

### CODE REVIEW

**Review by timurnkey — APPROVED** [review]
I like the direction 👍 

somewhat unrelated - I have some concerns about the `dist` directory bloating over time in `mono` with no structure. I _think_ in an ideal world we'd want a setup that looks like this:

```
# we already nested these under top-level `src` so I'd leave it for now
src/rust/<crate> # e.g. crate = notarizer/app
 - /src
 - /dist
src/go/<package> # e.g. package = tkcli
 - /cmd
 - /dist
```

I'm not sure how feasible this is OR if it adds too much complexity so LMK your thoughts

**Comment by lrvick** [comment]
@r-n-o All checks pass now, and there is now a working "make test" target. Any thoughts?

File: README.md
Comment (r-n-o) [inline README.md:35]: Is there a way to keep this brew path for install and verify the installed binaries as a post-install step? A lot of people won't have the patience to clone and verify, and I'm guessing a lot of people won't have gpg setup correctly or won't trust the signatures by default. I'm also not sure this flow is realistic for automated build pipelines (what if I simply want `turnkey` availble in my integration tests?)

It'd be nice to release the binaries and integrate with a package manager. I know you hate brew with a passion but that's what most of the devs on MacOSX use. Maybe we force another, more secure package manager on them? (MacPorts?)

File: README.md
Comment (r-n-o) [inline README.md:47]: We should be more explicit for macOS users and have a set of commands to copy/paste, ready to go. Bonus points for a cross-platform make target abstracting this (`make install`)

File: dist/manifest.36C8AAA9.asc
Comment (r-n-o) [inline dist/manifest.36C8AAA9.asc:1]: I don't see any docs on how this was generated, or how to verify this myself.  Guessing this is just an artifact of `make attest`?

File: README.md
Comment (r-n-o) [inline README.md:17]: Tried using this and ran into a bunch of issues:

1. Because of git submodules, `make attest` won't work. We need to make git submodule update/init part of the Makefile

2. The current makefile assumes users have `gpg` installed. This needs to be checked and a proper error message needs to be displayed if a user doesn't have `gpg` available or doesn't have the right version

3. When running `make verify` I see the following errors:
```
cat: cache/toolchain.state: No such file or directory
cat: cache/toolchain.state: No such file or directory
cat: cache/toolchain.state: No such file or directory

Verifying: dist/manifest.36C8AAA9.asc

gpg: Signature made Sat Mar 18 02:02:51 2023 CET
gpg:                using RSA key 67553FBDA46BB71ABD2E0B0B8E47A1EC35A1551D
gpg: Can't check signature: No public key
make: *** [verify] Error 2
```

I'm guessing `cache/toolchain.state: No such file or directory` isn't a fatal error, but we should make sure it's not present. We don't want to train users to ignore errors in `make verify`'s output.

The `Can't check signature: No public key` is the "real" error, not quite sure what triggers it.

4. Running `make attest` also yields a lot of `cat: cache/toolchain.state: No such file or directory` in the output.

5. Building (within `make attest`) took 152.8 seconds. This is a long time!

6. `make attest` failed with `groupadd: GID '20' already exists` (when trying to run `docker run --rm --tty --interactive --env UID=501 --env GID=20 --platform=linux/x86_64 --privileged --cpus 5 --volume /Users/rno/tkhq/code/tkcli:/home/build --workdir /home/build --env-file=config/global.env --env-file=cache/toolchain.env sha256:16bab33fc6ac040562cee29bc318881653a861d182253be2ec44a46d58aba140 src/toolchain/scripts/host-env bash -c ' GOHOSTOS="linux" GOHOSTARCH="amd64" GOOS="linux" GOARCH="386" GOCACHE=/home/build/cache/x86_64 GOPATH=/home/build/cache/x86_64 env -C src go build -o /home/build/out/turnkey.linux-386 main.go '`)

File: README.md
Comment (lrvick) [reply README.md:35]: > It'd be nice to release the binaries and integrate with a package manager. 

I fully agree, and hope we can see the turnkey binaries released in package managers that support reproducible builds and signing, such as Arch or Debian, but with good reason it takes time to earn a reputation and the trust of external maintainers to include it. We have to provide a best effort path for people with high security needs to form trust in binaries directly from us in advance of that.

> I know you hate brew with a passion but that's what most of the devs on MacOSX use. 

A lot of the security practices we provide like enclave based signing and webauthn request signing are not something most devs would know to ask for, as most devs lack a background in security. I have hope that if we are the ones that help people understand their threat model many will chose us over competing solutions that don't offer any options for high risk applications.

> Maybe we force another, more secure package manager on them? (MacPorts?)

Macports is slightly better, in that they bother to sign packages at all, though it is bulk automated signing so it is of limited value. The more important thing is they don't just let any randos push what they want like brew or npm, but the other edge of that sword is like linux distros you would have to go through the process to become a volunteer maintainer first and get the package approved which could prove challenging as it does not have a reputation others would recognize yet.

> I'm guessing a lot of people won't have gpg setup correctly or won't trust the signatures by default.

They should not trust the signatures by default, and that is a good thing. Encouraging people to stop and think about what they are doing is more important for their security than any engineering we can do.

gpg does ship with macos and virtually all linux distros. It also ships able to verify signatures with no configuration required.

> Is there a way to keep this brew path for install and verify the installed binaries as a post-install step?

I do not feel we should encourage use of unsigned package managers like brew on systems intended to participate in high accountability cryptography operations, which is all Turnkey use cases. That said if you feel the only way some people will kick the tires on Turnkey is with familiar low security paths, then a compromise would be providing such options with clear disclaimers.

We could then break up the installation paths broken out from easy but low security to difficult but high security.

Something like:

- low security
  - These options are provided only for convenience and familiarity
  - These options have no verification, so you could be downloading malware
  - We strongly encourage you to not use these methods on systems you need to be able to trust
  - brew ...
  - curl | bash ...
- medium security
  - Verify signatures on your binaries to ensure multiple people signed them
  - Verify the identities of some or all of the signers and that they are using well known keys
  - ...
- High security
  - Compile the binaries yourself and ensure they match the ones we publish
  - ...
  
If you agree with this path I can adjust the Readme along these lines, re-adding the brew install path as well as a bash one liner.

> I'm also not sure this flow is realistic for automated build pipelines (what if I simply want turnkey available in my integration tests?)

Automated build pipelines can generally git-clone anything to get easy access to the binaries in the dist directory, but the bash one liner could also exist for low security use cases like git hook integration tests.

File: dist/manifest.36C8AAA9.asc
Comment (r-n-o) [reply dist/manifest.36C8AAA9.asc:1]: Makes sense 👍

File: README.md
Comment (lrvick) [reply README.md:17]: > 1. Because of git submodules, make attest won't work. We need to make git submodule update/init part of the Makefile

It is in the makefile and automatically runs for me. If it does not for you let me know what version of GNU Make you are running. It would not shock me if MacOS uses a very old version we have to find workarounds for.

https://github.com/tkhq/tkcli/blob/lrvick/deterministic-builds/Makefile#L3-L5

> 2. The current makefile assumes users have gpg installed. This needs to be checked and a proper error message needs to be displayed if a user doesn't have gpg available or doesn't have the right version

I am not aware of any version of MacOS from the past 10 years or so, or any Linux distro, that does not ship with a version that can do "gpg --verify". If I am mistaken here lmk and I can address

> 3. When running make verify I see the following errors:

Fair. I'll add a make hook to fetch and pin all public keys for easy verification.

> 4. Running make attest also yields a lot of cat: cache/toolchain.state: No such file or directory in the output.

Fair. These were introduced in order to make mac builds run faster, but it is noisy. A workaround was not immediately obvious but I'll kick it out to a script if I have to.

> 5. Building (within make attest) took 152.8 seconds. This is a long time!

Make attest must do a from scratch build with no cache. This is just the time it takes to build the container image and run go build in it for each architecture with whatever number of cores you have assigned to Docker.

FWIIW it is only 70 seconds for me.

The same process for QoS takes 10-20m, so I consider this fast!

> 6. make attest failed with groupadd: GID '20' already exists

Ah. How annoying. MacOS apparently abuses user/group ids below 1000 which pretty much every other unix system reserves.  In this case the reserved gid '20' belongs to "dialout" in Linux. I can add logic to toolchain clobber any conflicting user/groups in the container at runtime.

File: README.md
Comment (r-n-o) [reply README.md:35]: > We could then break up the installation paths further from easy but low security to difficult but high security.

I like this 👍 

> Automated build pipelines can generally git-clone anything to get easy access to the binaries in the dist directory, but the bash one liner could also exist for low security use cases like git hook integration tests.

Makes sense. I also think we should provide a simple option to download and shasum check against known values we publish on our turnkey domain? We do this for e.g. awscli. Not the same as reproducible build, but it's a good option to ensure release binary integrity.

Comment (r-n-o) [reply README.md:35]: > They should not trust the signatures by default, and that is a good thing

Might be worth including steps to import keys in gpg such that make verify succeeds? I'm concerned about "expecting" `make verify` to fail. Users might think something's off with our releases instead of looking at the state of their keystore (see e.g. https://github.com/bitcoin-core/bitcoincore.org/issues/793)

Comment (lrvick) [reply README.md:35]: Make verify should succeed. I mean to say that it just will succeed but with keys with no trust history. We could output where people could go to verify the key fingerprints though.

> Makes sense. I also think we should provide a simple option to download and shasum check against known values we publish on our turnkey domain? We do this for e.g. awscli. Not the same as reproducible build, but it's a good option to ensure release binary integrity.

This means updating the website every time we make a new release. I would rather teach people to use and recognize that there should be multiple recognized signatures in the "make verify" output.

If we have a "team" page on the website, linking to the key fingerprints and keyoxide profiles of each engineer would allow people to form initial trust that given commit/release signature belong to members of our team

File: README.md
Comment (r-n-o) [reply README.md:17]: @lrvick I double-checked, and as far as I can tell `gpg` doesn't come pre-installed on MacOS. See https://support.apple.com/en-us/HT201214 and https://mikeross.xyz/gpg-without-gpgtools-on-mac/ for examples of install instructions

**Comment by lrvick** [comment]
@r-n-o Overhauled the ReadMe to reflect our conversation above.

If this all looks good I will make sure everything in it works as described. (Currently lies)

**Comment by r-n-o** [comment]
> @r-n-o Overhauled the ReadMe to reflect our conversation above.
> 
> If this all looks good I will make sure everything in it works as described. (Currently lies)

@lrvick I've taken a look just now: looks great! 🙌

File: README.md
Comment (r-n-o) [reply README.md:11]: (this is to make sure people who are linked straight to these low security instructions via deep links have a chance to "upgrade" to medium or high security if it's relevant / feasible for them)

**Review by jack-kearney — APPROVED** [review]

**Comment by r-n-o** [comment]
As far as Homebrew formulas go, I was able to install `dist/` binaries from v0.3.3 with the following formulae: https://github.com/r-n-o/homebrew-taptest/blob/459f4496be2f6634fce12f0dab625efaf38ef1ab/Formula/turnkey.rb#L1

Here's the output to prove it works!
```
$ brew tap r-n-o/taptest
==> Tapping r-n-o/taptest
Cloning into '/opt/homebrew/Library/Taps/r-n-o/homebrew-taptest'...
remote: Enumerating objects: 26, done.
remote: Counting objects: 100% (26/26), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 26 (delta 5), reused 26 (delta 5), pack-reused 0
Receiving objects: 100% (26/26), 7.64 KiB | 7.64 MiB/s, done.
Resolving deltas: 100% (5/5), done.
Tapped 1 formula (12 files, 13.1KB).

$ brew install r-n-o/taptest/turnkey
==> Fetching r-n-o/taptest/turnkey
==> Downloading https://github.com/tkhq/tkcli/raw/v0.3.3/dist/turnkey.darwin-aarch64
Already downloaded: /Users/rno/Library/Caches/Homebrew/downloads/7e8d66b5da53edf6e9f2ebc7f4a74478c59bd16398b5d59acc862fe87fca790b--turnkey.darwin-aarch64
==> Installing turnkey from r-n-o/taptest
🍺  /opt/homebrew/Cellar/turnkey/v0.3.3: 3 files, 8.8MB, built in 1 second
==> Running `brew cleanup turnkey`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).

$ which turnkey
/opt/homebrew/bin/turnkey

$ turnkey -h
NAME:
   tk - The Turnkey CLI

USAGE:
   tk [global options] command [command options] [arguments...]

COMMANDS:
   generate-api-key, gen     generate a new Turnkey API key
   request, r                make a request
   approve-request, approve  approve a request
   stamp, s                  sign an arbitrary message and produce a valid API Stamp
   help, h                   Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --help, -h  show help (default: false)
```

@lrvick

File: Makefile
Comment (r-n-o) [inline Makefile:48]: let's pull this make target out for now and bring the formula changes to [homebrew-tap](https://github.com/tkhq/homebrew-tap).

I'll spin up a PR to update the formula in that repo once this PR lands.

Comment (r-n-o) [reply Makefile:48]: Let's do this as a followup change

**Review by r-n-o — APPROVED** [review]

**Pull request merged**
